### PR TITLE
Firmware 102: A & B counter

### PR DESCRIPTION
**Pull request by tristanle22**
counterAB

### CODE REVIEW

File: projects/counter/src/main.c [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  LOG_DEBUG("Counter B: %d\n", counter->count_b);
+
+  soft_timer_start_millis(COUNTER_TIMEOUT_MILI_SECS, counter_a_timeout, counter, NULL);
+}
+
+int main(void) {
+  interrupt_init();
+  soft_timer_init();
+
+  Counters counter = {
```
Comment (g-jessmuir) [inline projects/counter/src/main.c:36]: please initialize this to zero with `Counters counter = { 0 }` to match convention

File: projects/counter/src/main.c [hunk 2]
```diff
@@ -0,0 +1,47 @@
+#include "counter.h"
+#include "interrupt.h"
+#include "log.h"
+#include "soft_timer.h"
+#include "wait.h"
+
+#define COUNTER_TIMEOUT_MILI_SECS 500
```
Comment (g-jessmuir) [inline projects/counter/src/main.c:7]: please just suffix the #define `_MS` to denote milli-seconds, to match with convention

File: projects/counter/inc/counter.h [hunk 3]
```diff
@@ -0,0 +1,10 @@
+#pragma once
+#include "interrupt.h"
+#include "log.h"
+#include "soft_timer.h"
+
+// Callback function for counter 1
+static void counter_a_timeout(SoftTimerId timer_id, void *context);
```
Comment (g-jessmuir) [inline projects/counter/inc/counter.h:7]: in general we'd consider these functions private and thus would not include them in the header. private functions are prefixed `prv_`, but the specs were not 100% clear on using only one file so this is ok. Functions in the header should not be declared static.

File: projects/counter/src/main.c [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  uint8_t count_a;
+  uint8_t count_b;
+} Counters;
+
+static void counter_a_timeout(SoftTimerId timer_id, void *context) {
+  Counters *counter = context;
+  counter->count_a++;
+  LOG_DEBUG("Counter A: %d\n", counter->count_a);
+
+  soft_timer_start_millis(COUNTER_TIMEOUT_MILI_SECS, counter_b_timeout, counter, NULL);
```
Comment (g-jessmuir) [inline projects/counter/src/main.c:19]: in the specs it mentioned using only one soft timer 🤔🤔🤔 I'll give you a pass here because it works though.

Comment (tristanle22) [reply projects/counter/src/main.c]: oh oops :sweat_smile:

**Comment by g-jessmuir** [comment]
complete!

**Pull request closed without merging**